File: BFS/16236번 아기 상어.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<algorithm>
#include<string.h>

#define MAX 20

using namespace std;

int dy[] = { 1, -1, 0, 0 };
int dx[] = { 0, 0, 1, -1 };

int n, ans = 0;
int pos_y, pos_x;
int current_size = 2, eat_cnt = 0;
int map[MAX][MAX];
int cnt[MAX][MAX];
vector<pair<int, pair<int, int>>> eat_fish;
queue<pair<int, int>> que;

void input()
{
	cin >> n;
	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < n; ++j) {
			cin >> map[i][j];

			if (map[i][j] == 9) {	// 처음 시작 위치 저장
				// 처음 시작 위치에 9가 저장되어 있으면 BFS를 수행할 때 물고기의 크기로 인식하고
				// 길을 지나지 못하는 경우가 발생하기 때문에 0으로 초기화 시켜줌
				map[i][j] = 0;
				pos_y = i;
				pos_x = j;
			}
		}
	}
}

void BFS(int y, int x)
{
	que.push({ y, x });

	while (!que.empty()) {
		int y = que.front().first;
		int x = que.front().second;
		que.pop();
		
		for (int i = 0; i < 4; ++i) {
			int ny = y + dy[i];
			int nx = x + dx[i];

			if (ny < 0 || ny >= n || nx < 0 || nx >= n) continue;

			// 한 번도 지나치지 않았으며 아기 상어가 지나갈 수 있는 길일 경우
			if (cnt[ny][nx] == 0 && current_size >= map[ny][nx]) {
				cnt[ny][nx] = cnt[y][x] + 1;

				// 해당 좌표에 물고기가 있으며 그 크기가 아기 상어보다 작을 경우
				if (map[ny][nx] != 0 && current_size > map[ny][nx])
						eat_fish.push_back({ cnt[ny][nx], {ny, nx} });

				que.push({ ny, nx });
			}
		}
	}
}

void init()
{
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
}

int main()
{
	init();
	input();
	
	while (1) {
		BFS(pos_y, pos_x);

		// 먹은 물고기가 없을 경우
		if (eat_fish.empty()) break;
		
		// 거리, y좌표, x좌표 순의 우선순위로 먹은 물고기를 정렬
		sort(eat_fish.begin(), eat_fish.end());

		ans += eat_fish[0].first;	// 가장 가까운 물고기의 거리 저장

		// 아기 상어의 위치를 먹은 물고기의 좌표로 변경
		pos_y = eat_fish[0].second.first;
		pos_x = eat_fish[0].second.second;

		map[pos_y][pos_x] = 0;	// 먹은 물고기 좌표를 0으로 초기화

		// 현재까지 먹은 물고기의 개수에 따라 아기 상어 크기 조절
		if (current_size == ++eat_cnt) {
			++current_size;
			eat_cnt = 0;
		}

		// 다음 사이클 돌기 전에 벡터, 배열 초기화
		eat_fish.clear();
		memset(cnt, 0, sizeof(cnt));
	}

	cout << ans;

	return 0;
}

File: 시뮬레이션/16236번 아기 상어.py
import sys
import heapq
from collections import deque

# 전처리 부분
input = sys.stdin.readline
dy = [1, -1, 0, 0]
dx = [0, 0, 1, -1]

# 알고리즘 부분
def bfs():
    eat_fish = []   # 아기 상어가 먹을 수 있는 물고기를 거리 기준으로 저장할 힙
    cnt = [[0] * n for _ in range(n)]   # 아기 상어가 이동한 거리를 저장하기 위한 배열

    que = deque([])
    que.append([pos_y, pos_x])

    while que:
        y, x = que.popleft()

        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]

            if 0 <= ny < n and 0 <= nx < n:
                if cnt[ny][nx] == 0 and size >= arr[ny][nx]:    # 이전에 지나간 칸이 아니며, 현재 사이즈가 해당 칸의 값보다 큰 경우
                    cnt[ny][nx] = cnt[y][x] + 1 # 이전 칸까지 이동한 거리 + 1

                    if arr[ny][nx] != 0 and size > arr[ny][nx]: # 해당 칸에 물고기가 존재하며 해당 물고기보다 아기 상어의 크기가 더 클 경우
                        heapq.heappush(eat_fish, [cnt[ny][nx], ny, nx]) # 문제에 제시된 조건에 따라 [거리, y축, x축] 순으로 힙에 저장

                    que.append([ny, nx])
    
    return eat_fish

# 메인 코드 부분
n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]
pos_y, pos_x = 0, 0
for i in range(n):
    for j in range(n):
        if arr[i][j] == 9:
            arr[i][j] = 0       # 0으로 초기화해주지 않으면 이후 bfs를 수행할 때 문제가됨
            pos_y, pos_x = i, j # 아기 상어의 초기 위치 저장

size = 2    # 아기 상어의 현재 사이즈
ans, eat_cnt = 0, 0
while True:
    eat_fish = bfs()    # 아기 상어가 잡아 먹을 수 있는 물고기 배열을 받아옴

    if not eat_fish: break  # 잡아 먹을 수 있는 물고기가 없는 경우

    cnt, pos_y, pos_x = eat_fish[0] # 힙에서 첫 번째 인덱스는 아기 상어가 먹을 수 있는 가장 가까운 물고기를 의미함
    ans += cnt      # 거리(시간) 더해주기
    eat_cnt += 1    # 물고기를 먹은 횟수 증가
    arr[pos_y][pos_x] = 0   # 물고기를 먹었으므로 해당 인덱스 0으로 초기화

    if size == eat_cnt: # 현재 아기 상어가 물고기를 먹은 횟수가 자신의 크기와 같은 경우
        size += 1       # 아기 상어 사이즈 증가
        eat_cnt = 0     # 횟수 초기화

print(ans)